python_modules/dagster/dagster/core/execution/plan/external_step.py

```diff
@@ -76,9 +77,13 @@ def launch_step(
         file_manager = LocalFileManager(".")
         events_file_handle = LocalFileHandle(events_file_path)
         events_data = file_manager.read_data(events_file_handle)
-        events = pickle.loads(events_data)
+        all_events = deserialize_value(pickle.loads(events_data))
 
-        yield from events
+        for event in all_events:
+            # write each pickled event from the external instance to the local instance
+            step_context.instance.handle_new_event(event)
+            if event.is_dagster_event:
+                yield event.dagster_event
```

Comment:
I had that same worry, but dagster events don't actually get stored in the instance by way of something consuming these generators and writing each event (part of the reason the original implementation had some weird bugs). DagsterEvents are added to the instance with functions like `log_step_event`, which actually get called when they are _instantiated_ (i.e. using `DagsterEvent.from_step(...)`). These remote events don't get added to the instance automatically because they are not instantiated using from_step() etc. methods locally (they're just deserialized).